worlds/terraria/__init__.py

```diff
@@ -0,0 +1,329 @@
+        for location in goal_locations:
+            _, flags, _, _ = rules[rule_indices[location]]
+            item = flags.get("Item") or f"Post-{location}"
+            ter_goals[item] = location
+            goal_items.add(item)
+
+        achievements = self.multiworld.achievements[self.player].value
+        location_count = 0
+        locations = []
+        for rule, flags, _, _ in rules[:goal]:
+            if (
+                (not self.calamity and "Calamity" in flags)
+                or (achievements < 1 and "Achievement" in flags)
+                or (achievements < 2 and "Grindy" in flags)
+                or (achievements < 3 and "Fishing" in flags)
+                or (
+                    rule == "Zenith" and self.multiworld.goal[self.player].value != 11
+                )  # Bad hardcoding
+            ):
+                continue
+            if "Location" in flags or ("Achievement" in flags and achievements >= 1):
+                # Location
+                location_count += 1
+                locations.append(rule)
+            elif (
+                "Achievement" not in flags
+                and "Location" not in flags
+                and "Item" not in flags
+            ):
+                # Event
+                locations.append(rule)
+
+        item_count = 0
+        items = []
+        for rule, flags, _, _ in rules[:goal]:
+            if not self.calamity and "Calamity" in flags:
+                continue
+            if "Item" in flags:
+                # Item
+                item_count += 1
+                if rule not in goal_locations:
+                    items.append(rule)
+            elif (
+                "Achievement" not in flags
+                and "Location" not in flags
+                and "Item" not in flags
+            ):
+                # Event
+                items.append(rule)
+
+        extra_checks = self.multiworld.fill_extra_checks_with[self.player].value
+        ordered_rewards = [
+            reward
+            for reward in labels["ordered"]
+            if self.calamity or "Calamity" not in rewards[reward]
+        ]
+        while extra_checks == 1 and item_count < location_count and ordered_rewards:
+            items.append(ordered_rewards.pop(0))
+            item_count += 1
+
+        random_rewards = [
+            reward
+            for reward in labels["random"]
+            if self.calamity or "Calamity" not in rewards[reward]
+        ]
+        self.multiworld.random.shuffle(random_rewards)
+        while extra_checks == 1 and item_count < location_count and random_rewards:
+            items.append(random_rewards.pop(0))
+            item_count += 1
+
+        while item_count < location_count:
+            items.append("Reward: Coins")
+            item_count += 1
+
+        self.ter_items = items
+        self.ter_locations = locations
+
+        self.ter_goals = ter_goals
+        self.goal_items = goal_items
+        self.goal_locations = goal_locations
```

Comment:
Define these fields in your class. It's bad practice to create new instance (or class) fields outside of `__init__`.

```python
class TerrariaWorld(World):
    # ...
    self.ter_items: <type>  # replace with your type
    self.ter_locations: <type>  # replace with your type
    self.ter_goals: <type>  # replace with your type
    self.goal_items: <type>  # replace with your type
    self.goal_locations: <type>  # replace with your type
```